### ISSUE TITLE
CrossAttention `causal_mask` depending on `attention_mask`.

### ISSUE THREAD

**Issue by gugarosa** [Dao-AILab/flash-attention#611]
Hello Tri Dao! First of all, thanks for all the research you have been conducting, it's amazing for the whole community.

Our group has some implementations based on Flash-Attention and we are implementing the Cross-Attention layer for speeding up decoding and supporting MQA/GQA.

However, I noticed that the base implementation at https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/modules/mha.py#L276 has this line:

```
sk = seqlen_k if key_padding_mask is None else rearrange(key_padding_mask.sum(-1), "b -> b 1 1 1")
```

Is it a typo? As far as I am concerned, the `causal_mask` should not rely on the attention/key padding mask. Based on the tests that I ran, using this line disrupts my outputs and make them far from `FlashCrossAttention` ones.

However, when ignoring it and just using the `seqlen_k` to produce the `causal_mask`, the results are comparable.

Could you please throw in some thoughts?

Thanks and best regards,
Gustavo.

**Comment by tridao** [comment]
It depends on how you define the causal mask.
Let say the key has length 3 [k0, k1, k2] but padded to [k0, k1, k2, pad].
And let's say you have 2 queries [q0, q1].
Then I think what that line of code is doing is constructing this mask:
       k0 k1 k2 pad
q0   1    1   0    0
q1    1    1   1    0
Do you have a different mask in mind?
